spackbot/pull_requests.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import contextlib
+import os
+import re
+import tempfile
+
+from sh.contrib import git
+from gidgethub import routing
+
+router = routing.Router()
+
+package_path = r"^var/spack/repos/builtin/packages/(\w[\w-]*)/package.py$"
+
+
+async def changed_packages(gh, repository, pull_request, number):
+    """Return an array of packages that were modified by a PR.
+
+    Ignore deleted packages, since we can no longer query them for
+    maintainers.
+
+    """
+    # see which files were modified
+    packages = []
+    async for f in gh.getiter(pull_request["url"] + "/files"):
+        filename = f["filename"]
+        status = f["status"]
+
+        if status == "removed":
+            continue
+
+        match = re.match(package_path, filename)
+        if not match:
+            continue
+        packages.append(match.group(1))
+
+    return packages
+
+
+def find_maintainers(repository, pull_request, number):
+    """Return an arrays of packages with maintainers, an array of packages
+    without maintainers, and a set of maintainers.
+
+    Ignore the author of the PR, as they don't need to review their own PR.
+    """
+    author = pull_request["user"]["login"]
+    clone_url = repository["clone_url"]
+
+    packages_with_maintainers = []
+    packages_without_maintainers = []
+    maintainers = set()
+
+    with temp_dir():
+        # Clone appropriate PR branch
+        git("clone", "#{clone_url}")
+        git("fetch", "origin", "pull/#{number}/head:PR#{number}", _cwd="spack")
+        git("checkout", "PR#{number}", _cwd="spack")
+
+        # Add `spack` to PATH
+        ENV["PATH"] = "#{dir}/spack/bin:#{ENV['PATH']}"
+
+        for package in packages:
+            logger.debug("Package: #{package}")
+
+            # Query maintainers
+            pkg_maintainers, status = Open3.capture2(
+                "spack", "maintainers", "#{package}"
+            )
+            pkg_maintainers = Set.new(pkg_maintainers.split)
+
+            logger.debug("Maintainers: #{pkg_maintainers}")
+
+            if pkg_maintainers.empty():
+                packages_without_maintainers.push(package)
+                continue
+
+            # No need to ask the author to review their own PR
+            pkg_maintainers.delete(author)
+
+            if not pkg_maintainers.empty:
+                packages_with_maintainers.push(package)
+                maintainers |= pkg_maintainers
+
+    return packages_with_maintainers, packages_without_maintainers, maintainers
+
+
+@router.register("pull_request", action="opened")
+@router.register("pull_request", action="reopened")
+@router.register("pull_request", action="closed")
```

Comment:
If I'm understanding this correctly, this will comment on PRs and ask maintainers to review when the PR is reopened or closed. I think we only want to comment on PRs once when a PR is first opened, otherwise we'll be commenting multiple times on the same PR.

Reply:
this was an artifact of testing